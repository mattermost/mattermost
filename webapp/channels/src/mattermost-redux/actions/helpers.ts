// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {ActionTypes} from 'mattermost-redux/action_types';
import {GenericAction} from 'mattermost-redux/types/actions';

export function getSharedChannels(): GenericAction {
    return {
        type: ActionTypes.GET_SHARED_CHANNELS,
    };
}

export function receivedSharedChannelsWithRemotes(sharedChannelsWithRemotes: any[]): GenericAction {
    return {
        type: ActionTypes.RECEIVED_SHARED_CHANNELS_WITH_REMOTES,
        data: sharedChannelsWithRemotes,
    };
}