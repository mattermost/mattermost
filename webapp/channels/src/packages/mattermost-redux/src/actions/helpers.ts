// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {AnyAction} from 'redux';

import type {ServerError} from '@mattermost/types/errors';
import type {GlobalState} from '@mattermost/types/store';

import {UserTypes} from 'mattermost-redux/action_types';
import {Client4} from 'mattermost-redux/client';
import type {DispatchFunc, GetStateFunc, ActionFuncAsync} from 'mattermost-redux/types/actions';

import {logError} from './errors';

type ActionType = string;
const HTTP_UNAUTHORIZED = 401;
export function forceLogoutIfNecessary(err: ServerError, dispatch: DispatchFunc<AnyAction>, getState: GetStateFunc) {
    const {currentUserId} = getState().entities.users;

    if ('status_code' in err && err.status_code === HTTP_UNAUTHORIZED && err.url && err.url.indexOf('/login') === -1 && currentUserId) {
        Client4.setToken('');
        dispatch({type: UserTypes.LOGOUT_SUCCESS, data: {}});
    }
}

function dispatcher(type: ActionType, data: any, dispatch: DispatchFunc<AnyAction>) {
    if (type.indexOf('SUCCESS') === -1) { // we don't want to pass the data for the request types
        dispatch(requestSuccess(type, data));
    } else {
        dispatch(requestData(type));
    }
}

export function requestData(type: ActionType) {
    return {
        type,
        data: null,
    };
}

export function requestSuccess(type: ActionType, data: any) {
    return {
        type,
        data,
    };
}

export function requestFailure(type: ActionType, error: ServerError): any {
    return {
        type,
        error,
    };
}

/**
 * Returns an ActionFunc which calls a specfied (client) function and
 * dispatches the specifed actions on request, success or failure.
 *
 * @export
 * @param {Object} obj                                       an object for destructirung required properties
 * @param {() => Promise<mixed>} obj.clientFunc              clientFunc to execute
 * @param {ActionType} obj.onRequest                         ActionType to dispatch on request
 * @param {(ActionType | Array<ActionType>)} obj.onSuccess   ActionType to dispatch on success
 * @param {ActionType} obj.onFailure                         ActionType to dispatch on failure
 * @param {...Array<any>} obj.params
 * @returns {ActionFunc} ActionFunc
 */

export function bindClientFunc<Func extends(...args: any[]) => Promise<any>>({
    clientFunc,
    onRequest,
    onSuccess,
    onFailure,
    params,
}: {
    clientFunc: Func;
    onRequest?: ActionType;
    onSuccess?: ActionType | ActionType[];
    onFailure?: ActionType;
    params?: Parameters<Func>;
}): ActionFuncAsync<Awaited<ReturnType<Func>>, GlobalState, AnyAction> {
    return async (dispatch, getState) => {
        if (onRequest) {
            dispatch(requestData(onRequest));
        }

        let data: Awaited<ReturnType<Func>>;
        try {
            if (params) {
                data = await clientFunc(...params);
            } else {
                data = await clientFunc();
            }
        } catch (error) {
            forceLogoutIfNecessary(error, dispatch, getState);
            if (onFailure) {
                dispatch(requestFailure(onFailure, error));
            }
            dispatch(logError(error));
            return {error};
        }

        if (Array.isArray(onSuccess)) {
            onSuccess.forEach((s) => {
                dispatcher(s, data, dispatch);
            });
        } else if (onSuccess) {
            dispatcher(onSuccess, data, dispatch);
        }

        return {data};
    };
}

// Debounce function based on underscores modified to use es6 and a cb

export function debounce(func: (...args: any) => unknown, wait: number, immediate?: boolean, cb?: () => unknown) {
    let timeout: NodeJS.Timeout|null;
    return function fx(...args: any[]) {
        const runLater = () => {
            timeout = null;
            if (!immediate) {
                Reflect.apply(func, null, args);
                if (cb) {
                    cb();
                }
            }
        };
        const callNow = immediate && !timeout;
        if (timeout) {
            clearTimeout(timeout);
        }
        timeout = setTimeout(runLater, wait);
        if (callNow) {
            Reflect.apply(func, null, args);
            if (cb) {
                cb();
            }
        }
    };
}

export class FormattedError extends Error {
    intl: {
        id: string;
        defaultMessage: string;
        values: any;
    };

    constructor(id: string, defaultMessage: string, values: any = {}) {
        super(defaultMessage);
        this.intl = {
            id,
            defaultMessage,
            values,
        };
    }
}

