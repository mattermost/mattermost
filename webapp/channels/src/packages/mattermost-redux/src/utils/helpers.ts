// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import shallowEqual from 'shallow-equals';

import {createSelectorCreator, defaultMemoize} from 'mattermost-redux/selectors/create_selector';

// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
export function memoizeResult<F extends Function>(func: F, measure: Function | undefined = undefined): F {
    let lastArgs: IArguments|null = null;
    let lastResult: any = null;

    // we reference arguments instead of spreading them for performance reasons
    return function memoizedFunc() {
        if (!shallowEqual(lastArgs, arguments)) { //eslint-disable-line prefer-rest-params
            //eslint-disable-line prefer-rest-params
            // apply arguments instead of spreading for performance.
            const result = Reflect.apply(func, null, arguments); //eslint-disable-line prefer-rest-params
            if (!shallowEqual(lastResult, result)) {
                lastResult = result;
            }
        }

        if (measure) {
            measure();
        }

        lastArgs = arguments; //eslint-disable-line prefer-rest-params
        return lastResult;
    } as unknown as F;
}

// Use this selector when you want a shallow comparison of the arguments and you want to memoize the result
// try and use this only when your selector returns an array of ids
export const createIdsSelector = createSelectorCreator(memoizeResult);

// Use this selector when you want a shallow comparison of the arguments and you don't need to memoize the result
export const createShallowSelector = createSelectorCreator(defaultMemoize, shallowEqual as any);

// isMinimumServerVersion will return true if currentVersion is equal to higher or than the
// the provided minimum version. A non-equal major version will ignore minor and dot
// versions, and a non-equal minor version will ignore dot version.
// currentVersion is a string, e.g '4.6.0'
// minMajorVersion, minMinorVersion, minDotVersion are integers
export const isMinimumServerVersion = (currentVersion: string, minMajorVersion = 0, minMinorVersion = 0, minDotVersion = 0): boolean => {
    if (!currentVersion || typeof currentVersion !== 'string') {
        return false;
    }

    const split = currentVersion.split('.');

    const major = parseInt(split[0], 10);
    const minor = parseInt(split[1] || '0', 10);
    const dot = parseInt(split[2] || '0', 10);

    if (major > minMajorVersion) {
        return true;
    }
    if (major < minMajorVersion) {
        return false;
    }

    // Major version is equal, check minor
    if (minor > minMinorVersion) {
        return true;
    }
    if (minor < minMinorVersion) {
        return false;
    }

    // Minor version is equal, check dot
    if (dot > minDotVersion) {
        return true;
    }
    if (dot < minDotVersion) {
        return false;
    }

    // Dot version is equal
    return true;
};

// Generates a RFC-4122 version 4 compliant globally unique identifier.
export function generateId(): string {
    // implementation taken from http://stackoverflow.com/a/2117523
    let id = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx';
    id = id.replace(/[xy]/g, (c) => {
        const r = Math.floor(Math.random() * 16);
        let v;

        if (c === 'x') {
            v = r;
        } else {
            // eslint-disable-next-line no-mixed-operators
            v = r & 0x3 | 0x8;
        }

        return v.toString(16);
    });
    return id;
}

export function isEmail(email: string): boolean {
    // writing a regex to match all valid email addresses is really, really hard. (see http://stackoverflow.com/a/201378)
    // this regex ensures:
    // - at least one character that is not a space, comma, or @ symbol
    // - followed by a single @ symbol
    // - followed by at least one character that is not a space, comma, or @ symbol
    // this prevents <Outlook Style> outlook.style@domain.com addresses and multiple comma-separated addresses from being accepted
    return (/^[^ ,@]+@[^ ,@]+$/).test(email);
}
