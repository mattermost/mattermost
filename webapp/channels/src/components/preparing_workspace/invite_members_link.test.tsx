// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {withIntl} from 'tests/helpers/intl-test-helper';
import {fireEvent, render, screen} from 'tests/react_testing_utils';

import InviteMembersLink from './invite_members_link';

describe('components/preparing-workspace/invite_members_link', () => {
    const inviteURL = 'https://invite-url.mattermost.com';

    it('should match snapshot', () => {
        const component = withIntl(<InviteMembersLink inviteURL={inviteURL}/>);

        const {container} = render(component);
        expect(container).toMatchSnapshot();
    });

    it('should match snapshot when displayed including the input field', () => {
        const component = withIntl(
            <InviteMembersLink
                inviteURL={inviteURL}
                inputAndButtonStyle={true}
            />,
        );

        const {container} = render(component);
        expect(container).toMatchSnapshot();
    });

    it('renders only with the button if the inputAndButton option is false', () => {
        const component = withIntl(
            <InviteMembersLink
                inviteURL={inviteURL}
                inputAndButtonStyle={false}
            />,
        );
        render(component);
        const input = screen.queryByText(inviteURL);
        expect(input).not.toBeInTheDocument();
        const button = screen.getByRole('button', {name: /copy link/i});
        expect(button).toBeInTheDocument();
    });

    it('renders an input field with the invite URL', () => {
        const component = withIntl(
            <InviteMembersLink
                inviteURL={inviteURL}
                inputAndButtonStyle={true}
            />,
        );
        render(component);
        const input = screen.getByDisplayValue(inviteURL);
        expect(input).toBeInTheDocument();
    });

    it('renders a button to copy the invite URL', () => {
        const component = withIntl(<InviteMembersLink inviteURL={inviteURL}/>);
        render(component);
        const button = screen.getByRole('button', {name: /copy link/i});
        expect(button).toBeInTheDocument();
    });

    it('changes the button text to "Link Copied" when the URL is copied', () => {
        const component = withIntl(<InviteMembersLink inviteURL={inviteURL}/>);
        render(component);
        const button = screen.getByRole('button', {name: /copy link/i});
        const originalText = 'Copy Link';
        const linkCopiedText = 'Link Copied';
        expect(button).toHaveTextContent(originalText);

        fireEvent.click(button);

        expect(button).toHaveTextContent(linkCopiedText);
    });
});
