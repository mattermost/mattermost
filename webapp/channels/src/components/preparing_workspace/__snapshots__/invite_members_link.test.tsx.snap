// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`components/preparing-workspace/invite_members_link should match snapshot 1`] = `
<div>
  <div
    class="InviteMembersLink"
  >
    <button
      class="InviteMembersLink__button--single"
      data-testid="shareLinkInputButton"
    >
      <i
        class="icon icon-content-copy"
      />
      <span>
        Copy Link
      </span>
    </button>
  </div>
</div>
`;

exports[`components/preparing-workspace/invite_members_link should match snapshot when displayed including the input field 1`] = `
<div>
  <div
    class="InviteMembersLink"
  >
    <input
      aria-label="team invite link"
      class="InviteMembersLink__input"
      data-testid="shareLinkInput"
      readonly=""
      type="text"
      value="https://invite-url.mattermost.com"
    />
    <button
      class="InviteMembersLink__button"
      data-testid="shareLinkInputButton"
    >
      <i
        class="icon icon-link-variant"
      />
      <span>
        Copy Link
      </span>
    </button>
  </div>
</div>
`;
